### ISSUE TITLE
Adopt CSI for Azure extension beginning with Kubernetes 1.21

### ISSUE THREAD

**Issue by rfranzke** [gardener/gardener-extension-provider-azure#3]
As of Kubernetes 1.18 the in-tree Azure Disk volume driver is deprecated in favour of CSI. We should adopt CSI for 1.18+ clusters (where the Azure Disk migration will be promoted to beta as well)

See also: https://github.com/kubernetes/kubernetes/pull/85231 and https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/

**Comment by rfranzke** [comment]
The CSI migration is beta with 1.17, though, the CSI plugins for Azure are not yet ready to cater with Gardener needs. We will have to contribute there before we can migrate to CSI. Hence, postponing this issue until 1.19.

**Comment by rfranzke** [comment]
I opened https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/354 for the Azure Disk CSI plugin. After it's merged I'll use the same approach to enhance the [Azure File CSI plugin](https://github.com/kubernetes-sigs/azurefile-csi-driver).

**Comment by rfranzke** [comment]
Turns out that what I described with https://github.com/gardener/gardener-extension-provider-azure/issues/3#issuecomment-610922049 was actually a misconfiguration during my tests - the Azure Disk CSI plugin already works for our scenario as it disables the metadata service in the controller service of the CSI driver. Hence https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/354 is closed again for now.

However, it seems that the node service of the CSI driver also requires the client id/secret credentials in order to talk to the Azure API and find out information about certain disks. This would mean that we would have to expose the credentials in the shoot cluster which is undesired. I'm investigating further...

**Comment by rfranzke** [comment]
I checked a little bit more, and it seems that the [`checkDiskExists()`](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/v0.6.0/pkg/azuredisk/nodeserver.go#L331-L333) function in the `NodeGetVolumeStats` function is the only place in the node service where the credentials are needed right now.
The Azure CSI maintainers offered to look into whether it is possible to get rid of this check. Afterwards it should be possible to start the CSI driver without client id/secret.

**Comment by rfranzke** [comment]
I played with the latest [`kubernetes-sigs/azuredisk-csi-driver@v0.7.0`](https://github.com/kubernetes-sigs/azuredisk-csi-driver/releases/tag/v0.7.0) and [`kubernetes-sigs/azurefile-csi-driver@v0.6.0`](https://github.com/kubernetes-sigs/azurefile-csi-driver/releases/tag/v0.6.0) releases and it seems that the problems we faced have been resolved. This would mean that we can go ahead and introduce CSI (migration) already with 1.18 as initially planned.
I'll do some more testing tomorrow, and if successful I'll add the respective changes to the currently open 1.18 support PR #72. I'll also update this ticket with my results.

**Comment by rfranzke** [comment]
It turns out that the Azure CSI migration is still alpha with 1.18 (although suggested differently by https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/), hence, let's plan for 1.19. There are still flaws in the Azure File CSI migration, the Azure Disk CSI migration worked so far.
See also https://github.com/kubernetes/enhancements/issues/1490.

**Comment by rfranzke** [comment]
/reopen
We decided to postpone CSI migration for Azure to Kubernetes v1.21, so let's adapt the version constants in the code from 1.20 to 1.21.

**Comment by ialidzhikov** [comment]
> We decided to postpone CSI migration for Azure to Kubernetes v1.21, so let's adapt the version constants in the code from 1.20 to 1.21.

Let's wait few more days to make sure that `CSIMigrationAzureFile` promotion to beta will be pushed back to v1.21 - ref https://github.com/kubernetes/kubernetes/pull/96293. Theoretically it can get in for v1.20 via an [Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md).
